core/src/main/scala/org/apache/spark/storage/BlockManager.scala

```diff
@@ -1424,6 +1457,16 @@ private[spark] class BlockManager(
     blockStoreUpdater.save()
   }
 
+  // Check whether a rdd block is visible or not.
+  private[spark] def isRDDBlockVisible(blockId: RDDBlockId): Boolean = {
+    // If the rdd block visibility information not available in the block manager,
+    // asking master for the information.
+    if (blockInfoManager.isRDDBlockVisible(blockId)) {
+      return true
+    }
+    master.isRDDBlockVisible(blockId)
```

Comment:
QQ: Should we be caching this instead of hitting driver each time ?

Reply:
Yes, I've thought about this. Then we will add a mechanism to clean the cache then.

Reply:
One way in my mind is that, we cache the results(for visible rdd blocks) in the block manager. Once the rdd is removed, a broadcast message will be sent to each BlockManager to clean the cache.

I am wondering is it worth to do this? Since there is locality scheduling, if tasks got scheduled to the executor where cached block exists, there'll be no calls to master.

Let me know your thoughts about this, thanks. cc @mridulm @Ngone51

Reply:
> Once the rdd is removed, a broadcast message will be sent to each BlockManager to clean the cache.

I think the problem in this way is that it introduces the race condition between the visible cache and the broadcast message.

`master.isRDDBlockVisible(blockId)` shouldn't be frequently called, right? I thought `blockInfoManager.isRDDBlockVisible(blockId)` would be the most case for us, no?

Reply:
What I meant was, cache visibility - once a block becomes visible, we do not revert it back to being invisible.

Reply:
> > Once the rdd is removed, a broadcast message will be sent to each BlockManager to clean the cache.
> 
> I think the problem in this way is that it introduces the race condition between the visible cache and the broadcast message.
> 
> `master.isRDDBlockVisible(blockId)` shouldn't be frequently called, right? I thought `blockInfoManager.isRDDBlockVisible(blockId)` would be the most case for us, no?

With locality scheduler I think `master.isRDDBlockVisible(blockId) ` should not be frequently called. If the assumption is true, then maybe we don't need to cache the results.

Reply:
To clarify, (my previous comment was poorly worded).

A block would be invisible for three reasons primarily (and minor variants of these) :
a) Some task T1 is still actively working on the partition which computed it.
b) Some task T1 on E1 generated the partition, which got pushed to E2 due to replication - and is not yet visible, since T1 is still running.
c) Some task T1 on E1 generated B1, but failed - and so block exists on E1 but is invisible.


In these cases, a block which is invisible would become visible once the task T1 successfully complete - or (in case of failure), is dropped and becomes invisible.
So the transition is only from invisible -> visible or invisible -> dropped.
When a block transitions to becoming visible (due to an explicit message from driver to the executor which generated it, and to the replicas), it remains in that state until it is dropped explicitly (which is again another message from driver).

So the proposal was:
a) Maintain cached state of block at executor - once it is visible, it never changes from that.
b) If invisible, it could be a potential race (T1 completed, driver yet to update B1 to E1, but launched some T2 which wants to read B1) - so we have to query driver, but we can cache the result is visible.
b.1) We can minimize this impact by dropping all B1's when the task T1 fails at E1 (driver would send the update anyway).


So essentially, we can cache only if the block is visible - if it is not visible, we always have to query driver (with (b.1) to minimize needing to query failures)

Thoughts ?

Reply:
Hi @mridulm, in current implementation, once a block turns to be visible, driver would send a broadcast message to executors having the cached block data stored to mark the block as visible.

The state `visibleRDDBlocks` is cached in [BlockerInfoManager](https://github.com/apache/spark/pull/39459/files#diff-fdee2ef66ad5bea5323506395b453145c74f47c8da092dcacd34a66190a20a15). It is kind of cached visiblity state in executor side but only in executors which have the cached block stored. This is done in a push-based update style.

With above mechanism, do you think we still need another cache to store the visiblity information in executor or do we also need to cache the state in executors not having the cached block data stored?

Reply:
> T1 completed, driver yet to update B1 to E1, but launched some T2 which wants to read B1

I'm not trying to talk about the cache stuff. But just try to highlight this. So this should somehow be a behavior change, right? If T1 generates B1 in early time and T1 turns out to be a long-running task, it can be terrible for tasks like T2 which reads B1.

Reply:
> With above mechanism, do you think we still need another cache to store the visiblity information in executor or do we also need to cache the state in executors not having the cached block data stored?

You are right, we dont need it: since we are already checking for `blockInfoManager.isRDDBlockVisible(blockId)`. Agree

Reply:
> I'm not trying to talk about the cache stuff. But just try to highlight this. So this should somehow be a behavior change, right? If T1 generates B1 in early time and T1 turns out to be a long-running task, it can be terrible for tasks like T2 which reads B1.

Do we actually have usecases where some other stage/task is depending on an earlier task generating a block as a prefix to its computation ?

Looks like a fairly brittle assumption, no ? (Unless I misunderstood the usecase here !)

Reply:
Yes, `blockInfoManager.isRDDBlockVisible(blockId)` should check for the current executor side state of the block to see if it is visible or not.
If visible, immediately return true.
Else, query driver for visibility status - if driver returns true, cache that.

Also, when a block becomes visible at driver, it will update the block managers hosting the block about the visibility status - which is then maintained at executor.

Reply:
Updated, please take a look. Thanks.